systemd/nslogin/nslogin.c

```diff
@@ -0,0 +1,315 @@
+    printf("euid: %d\n", geteuid());
+    printf("uid: %d\n", getuid());
+#endif
+    // Wait for systemd to be ready
+    pid_t systemdPid = find_systemd();
+    if (systemdPid == 0) {
+        fprintf(stderr, "Systemd is not running. " RESTART_MSG);
+        exit(1);
+    }
+    if (!wait_on_systemd()) {
+        fprintf(stderr, RESTART_MSG);
+        exit(1);
+    }
+
+    // Gain root privilege
+    if (setegid(0) != 0) {
+        perror("setegid as root");
+        exit(2);
+    }
+    if (seteuid(0) != 0) {
+        perror("seteuid as root");
+        exit(3);
+    }
+
+    // Enter systemd namespace
+    if (enter_target_ns(systemdPid) != 0) {
+        perror("Failed to set namespace to systemd's");
+        exit(4);
+    }
+
+    // Drop privileges back as the current user
+    if (seteuid(uid) != 0) {
+        perror("seteuid back as user");
+        exit(5);
+    }
+    if (setegid(gid) != 0) {
+        perror("setegid back as user");
+        exit(6);
+    }
+    // Make sure any pending output is flushed before forking.
+    fflush(stdout);
+    fflush(stderr);
+
+    // Without forking after setting PID namespace, we get setpgid errors when piping commands.
+    //  child setpgid (PID_X to PID_Y): Operation not permitted
+    continue_as_child();
+    // Having no argument means that we want to shell in.
+    int argvslen = 1;
+    if (argc > 1) {
+        argvslen = argc - 1; // argv[0] is this binary, skip it.
+    }
+    char *argvs[argvslen + 1];
+    argvs[argvslen] = NULL;
+
+    // Entering shell as default behaviour.
+    char *pathname = getenv("SHELL");
+    argvs[0] = pathname;
+
+    // We pointed to a binary to execute instead of entering the default shell.
+    if (argc > 1) {
+        pathname = argv[1]; // executable is passed as first argument
+        for (int i = 1; i < argc; ++i) {
+            argvs[i - 1] = argv[i];
+        }
+    }
+
+    // Finally exec.
+    if (execvp(pathname, argvs) == -1) {
+        const size_t size = 128;
+        char msg[size];
+        snprintf(msg, size, "Executing \"%s\" failed", pathname);
+        perror(msg);
+        exit(7);
+    }
+} // main.
+
+/// Returns true if systemd completed its startup.
+static bool is_systemd_startup_complete(sd_bus *bus) {
+    char *msg = NULL;
+    // sd_bus_error_free is safe to call even if there is no error data.
+    sd_bus_error err __attribute__((__cleanup__(sd_bus_error_free))) = SD_BUS_ERROR_NULL;
+    int busOk = sd_bus_get_property_string(bus, "org.freedesktop.systemd1", "/org/freedesktop/systemd1",
+                                           "org.freedesktop.systemd1.Manager", "SystemState", &err, &msg);
+    if (busOk < 0) {
+        fprintf(stderr, "Bus error %s", err.message);
+        return false;
+    }
+
+    if (msg == NULL) {
+        return false;
+    }
+    // We won't get the message "Failed to connect to bus: No such file or directory" at this point because there is a
+    // bus available at this point.
```

Comment:
And anyway, I think this would trigger `busOK <0`.

Reply:
Exactly. That causes an early return, so we never reach the string comparation line. Thus no need to compare msg for "Failed to connect to bus..." message.